Lib/_pyrepl/utils.py

```diff
@@ -41,25 +78,215 @@ def unbracket(s: str, including_content: bool = False) -> str:
     return s.translate(ZERO_WIDTH_TRANS)
 
 
-def disp_str(buffer: str) -> tuple[CharBuffer, CharWidths]:
-    r"""Decompose the input buffer into a printable variant.
+def gen_colors(buffer: str) -> Iterator[ColorSpan]:
+    """Returns a list of index spans to color using the given color tag.
+
+    The input `buffer` should be a valid start of a Python code block, i.e.
+    it cannot be a block starting in the middle of a multiline string.
+    """
+    sio = StringIO(buffer)
+    line_lengths = [0] + [len(line) for line in sio.readlines()]
+    # make line_lengths cumulative
+    for i in range(1, len(line_lengths)):
+        line_lengths[i] += line_lengths[i-1]
+
+    sio.seek(0)
+    gen = tokenize.generate_tokens(sio.readline)
+    last_emitted: ColorSpan | None = None
+    try:
+        for color in gen_colors_from_token_stream(gen, line_lengths):
+            yield color
+            last_emitted = color
+    except tokenize.TokenError as te:
+        yield from recover_unterminated_string(
+            te, line_lengths, last_emitted, buffer
+        )
+
+
+def recover_unterminated_string(
+    exc: tokenize.TokenError,
+    line_lengths: list[int],
+    last_emitted: ColorSpan | None,
+    buffer: str,
+) -> Iterator[ColorSpan]:
+    msg, loc = exc.args
+    if (
+        msg.startswith("unterminated string literal")
+        or msg.startswith("unterminated f-string literal")
+        or msg.startswith("EOF in multi-line string")
+        or msg.startswith("unterminated triple-quoted f-string literal")
+    ):
+        start = line_lengths[loc[0] - 1] + loc[1] - 1
+        end = line_lengths[-1] - 1
+
+        # in case FSTRING_START was already emitted
+        if last_emitted and start <= last_emitted.span.start:
+            trace("before last emitted = {s}", s=start)
+            start = last_emitted.span.end + 1
+
+        span = Span(start, end)
+        trace("yielding span {a} -> {b}", a=span.start, b=span.end)
+        yield ColorSpan(span, "STRING")
+    else:
+        trace(
+            "unhandled token error({buffer}) = {te}",
+            buffer=repr(buffer),
+            te=str(exc),
+        )
+
+
+def gen_colors_from_token_stream(
+    token_generator: Iterator[TI],
+    line_lengths: list[int],
+) -> Iterator[ColorSpan]:
+    token_window = prev_next_window(token_generator)
+
+    is_def_name = False
+    paren_level = 0
+    for prev_token, token, next_token in token_window:
+        assert token is not None
+        if token.start == token.end:
+            continue
+
+        match token.type:
+            case T.STRING | T.FSTRING_START | T.FSTRING_MIDDLE | T.FSTRING_END:
```

Comment:
Since we're targeting 3.14, we might also want to support t-strings

Reply:
Show me a TSTRING token and I will add it.

Reply:
This was added yesterday in https://github.com/python/cpython/pull/132662/files#diff-62d6ae4a2b687e27352e989df2b323cee846a5f6d00a28e9705ba410ced1200cR71-R73:

```python
~ echo "t''" | ./python -m tokenize          
1,0-1,2:            TSTRING_START  "t'"           
1,2-1,3:            TSTRING_END    "'"            
1,3-1,4:            NEWLINE        '\n'           
2,0-2,0:            ENDMARKER      ''
```

Reply:
Alright, adding. Thanks!